network_simulator/plot_tool.py

```diff
@@ -0,0 +1,300 @@
+        output = []
+        for tup_idx in range(len(tuple_list)):
+            timestamp, data = tuple_list[tup_idx]
+            if timestamp > curr_time:
+                if curr_count != 0:
+                    output.append((curr_time, float(curr_sum) / curr_count))
+                    curr_count = 0
+                    curr_sum = 0
+                curr_time += GRAPH_WINDOW_SIZE
+            if timestamp <= curr_time:
+                curr_count += 1
+                curr_sum += data
+        # Append the last elements if they did not reach an end of an
+        # interval.
+        if curr_count != 0:
+            output.append((curr_time, float(curr_sum) / curr_count))
+        return output
+
+    def graph_tuple_list(self, tuple_list=[], scatter=False):
+        """
+        Graphs the content of a tuple list.
+
+        :param list tuple_list: if we have a newer version of the list,
+        we would like to replace the previous version.
+        :param boolean scatter: if we want a scatter plot, we set this
+        variable to True; otherwise, it would be a line graph.
+        """
+        if not tuple_list:
+            return
+
+        if scatter:
+            # Scatter plot.
+            plt.scatter(*zip(*tuple_list), marker='o', linestyle='--')
+        else:
+            # Line graph
+            plt.plot(*zip(*tuple_list), marker='o', linestyle='--')
+
+    def graph_1d_list(self, lst=[]):
+        """
+        Graphs the content of a list.
+
+        :param list lst: list of values for the 1-D graph.
+        """
+        if not lst:
+            return
+
+        # Plot 1-D different values on a x-y plane.
+        plt.plot(lst, np.zeros_like(lst), marker='o', linestyle='--')
+
+class Grapher(object):
+    """
+    Graphs statistics items from Statistics class.
+    """
+    def __init__(self, stats=None):
+        """
+        :ivar Statistics stats: input Statistics object for graphing.
+        :ivar PlotTool tool: graphing functions.
+        """
+        self.stats = stats
+        self.tool = PlotTool()
+
+    def graph_links(self):
+        """
+        Graphs the link stats of each available link.
+        """
+        # Set up legend storage for link names.
+        occpy_legend = []
+        loss_legend = []
+        trans_legend = []
+
+        # Iterate through each link to display each stats.
+        for link, link_stats in self.stats.link_stats.iteritems():
+            # Name of link would be "link_name".
+            # Buffer occupancy w.r.t time.
+            plt.subplot(431)
+            self.tool.graph_tuple_list(link_stats.buffer_occupancy)
+            occpy_legend.append(link)
+
+            # Packet loss times.
+            plt.subplot(434)
+            loss_rate_lst = self.tool.gen_count_interval_list(
+                link_stats.packet_loss_times)
+            self.tool.graph_tuple_list(loss_rate_lst)
+            loss_legend.append(link)
+
+            # Packet transmission w.r.t time.
```

Comment:
Don't graph packet transmission vs time. What we want for the third graph is the link flow rate vs time. This is a rate that must be calculated through windowing.